### PR TITLE
Tokenizer

### PR DESCRIPTION
**Pull request by DarkMarmot**
This adds a fast path for HL7 messages with the default delimiters by using a tokenizer path (instead String splitting) and then converting the tokens to nested lists. HL7 parsing speed increases by nearly a factor of 3!

### CODE REVIEW

File: lib/hl7/parser.ex [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  @repetition 2
+  @component 1
+  @sub_component 0
+
+  def parse(<<"MSH|^~\\&", rest::binary()>> = text) do
+    tokenize(rest, text, 8, 0, ["^~\\&", @field, "|", @field, "MSH"])
+    |> to_lists()
+  end
+
+  def tokenize(<<char::binary-size(1), rest::binary>>, original, skip, len, acc) do
```
Comment (calvinb) [inline lib/hl7/parser.ex:13]: Do all of these functions need to be public?

File: lib/hl7/parser.ex [hunk 2]
```diff
@@ -0,0 +1,118 @@
+defmodule HL7.Parser do
+  @carriage_return 4
+  @field 3
+  @repetition 2
+  @component 1
+  @sub_component 0
+
+  def parse(<<"MSH|^~\\&", rest::binary()>> = text) do
```
Comment (calvinb) [inline lib/hl7/parser.ex:8]: Are there any tests that call this function?

File: lib/hl7/parser.ex (hunk 1 above)
Comment (calvinb) [reply lib/hl7/parser.ex:13]: Curious why we preserve `original` and pass it around with a `skip` instead of chopping it down as we progress? (Which would obviously lead to a name change.)

File: lib/hl7/parser.ex (hunk 2 above)
Comment (calvinb) [reply lib/hl7/parser.ex:8]: I'm writing some. Maybe I'll PR them to your branch.

File: lib/hl7/message.ex [hunk 3]
```diff
@@ -131,6 +131,20 @@ defmodule HL7.Message do
     msg
   end
 
+  def new(<<"MSH|^~\\&", _rest::binary()>> = raw_text) do
```
Comment (calvinb) [inline lib/hl7/message.ex:134]: Is it okay to assume the input is using the standard delimiters?

File: lib/hl7/parser.ex (hunk 2 above)
Comment (DarkMarmot) [reply lib/hl7/parser.ex:8]: It got test indirectly (as almost every existing test went through this path now....) but I added an explicit test that comparse the list parse of the original code and this new code to verify that they are identical.

File: lib/hl7/message.ex (hunk 3 above)
Comment (DarkMarmot) [reply lib/hl7/message.ex:134]: yup :) The parser code is only used for default limiters (as with this case). I found that I lost the performance benefits if I made them dynamic and used pinned case stuff to split based on the separators in the message. I've never seen an HL7 message that didn't use the defaults, though, so having this high performance path seems like a good way to go.

Comment (DarkMarmot) [reply lib/hl7/message.ex:134]: If the separators are not the defaults, it's still using the old code.

File: lib/hl7/parser.ex (hunk 1 above)
Comment (DarkMarmot) [reply lib/hl7/parser.ex]: The `original` is used to grab strings out (that preceded where we are in the parse). This is code I blatantly ripped off from the Jason implementation ;)

Comment (calvinb) [reply lib/hl7/parser.ex]: I looked for something like that, but I'm still missing it. Where does that happen?

File: test/hl7_message_test.exs [hunk 4]
```diff
@@ -72,7 +72,7 @@ defmodule HL7MessageTest do
     new_msg = raw_text_with_garbage |> HL7.Message.new()
     roundtrip = new_msg |> to_string()
     assert roundtrip == raw_text
-    assert new_msg.fragments == ["garbage text"]
+    assert new_msg.fragments == [["garbage text"]]
```
Comment (calvinb) [inline test/hl7_message_test.exs:75]: Curious why this changed.

File: lib/hl7/parser.ex [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+  defp split_by(tokens, delimiter) do
+    split_by(tokens, delimiter, [], [])
+  end
+
+  defp split_by([], _delimiter, buffer, result) do
+    Enum.reverse([Enum.reverse(buffer) | result])
+  end
+
+  defp split_by([delimiter | rest], delimiter, buffer, result) do
```
Comment (calvinb) [inline lib/hl7/parser.ex:109]: This is very cool, but it took me a minute to see how it would ever fail to match a call that would match the next head. I can't think of a way to draw any more attention to `delimiter`; the names are good.

File: test/hl7_message_test.exs (hunk 4 above)
Comment (DarkMarmot) [reply test/hl7_message_test.exs:75]: It was how I was chucking unparseable HL7 junk off to the side (which ended up being slightly different in the new version). I might try to figure out a better way to handle it -- but I'm not really sure this test even makes sense as is (since neither version keeps the garbage in the right place I noticed if you want a roundtrip, just appends it at the end). I would need to make structural changes to the library to support more things -- and I wanted to minimize my changes here for safety.

File: lib/hl7/parser.ex (hunk 1 above)
Comment (DarkMarmot) [reply lib/hl7/parser.ex]:
```
 defp tokenize_terminator(text, original, skip, len, acc, terminator) do
    string = binary_part(original, skip, len)
    tokenize(text, original, skip + len + 1, 0, [terminator, string | acc])
  end
  ```
The 'binary_part` call pulls it out (we keep incrementing skip and len as we go through the text).

File: lib/hl7/parser.ex (hunk 5 above)
Comment (DarkMarmot) [reply lib/hl7/parser.ex:109]: Thanks :) Yeah, I didn't know of a nice way to make this kind of recursive thing easy to read. 
It's basically `String.split` but for `Enum` -- but Elixir already has that function and it doesn't match my expectations of acting like it does in `String` :(

File: lib/hl7/parser.ex (hunk 1 above)
Comment (calvinb) [reply lib/hl7/parser.ex]: I don't see `original` referenced anywhere without `binary_part` and `skip`. Why not chop its head off instead of incrementing `skip` and passing it around as a marker of where to start?

File: lib/hl7/parser.ex (hunk 5 above)
Comment (calvinb) [reply lib/hl7/parser.ex:109]: I think this will do the same thing but a bit less to look at:
```elixir
  defp split_by(tokens, delimiter) do
    List.foldr(tokens, [[]], fn
      ^delimiter, [head | tail] -> [[] | [head | tail]]
      token, [head | tail] -> [[token | head] | tail]
    end)
  end
```

File: lib/hl7/parser.ex [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+  @repetition 2
+  @component 1
+  @sub_component 0
+
+  def parse(<<"MSH|^~\\&", rest::binary()>> = text) do
+    tokenize(rest, text, 8, 0, ["^~\\&", @field, "|", @field, "MSH"])
+    |> to_lists()
+  end
+
+  defp tokenize(<<char::binary-size(1), rest::binary>>, original, skip, len, acc) do
```
Comment (calvinb) [inline lib/hl7/parser.ex:13]: I'm finding the indirect recursion between `tokenize` and `tokenize_terminator` a little hard to follow. I was thinking you could probably do the same work with a fold over the characters and no explicit recursion. I had to try it before I suggested it, and it seems to work for me. I know the focus of this PR is performance, which might be affected by any tweak to the approach, so I'm going to try to get `Benchee` results on the fold.

Comment (calvinb) [reply lib/hl7/parser.ex:13]: Argh, the fold made it slower! My world is shattered. Who cares about performance anyway?

Comment (calvinb) [reply lib/hl7/parser.ex:13]: I took a shot at inlining `tokenize_terminator` to make the recursion easier to follow. See what you think about this.
```elixir
  defp tokenize(<<char::binary-size(1), rest::binary>>, original, skip, len, acc) do
    case get_terminator(char) do
      nil ->
        tokenize(rest, original, skip, len + 1, acc)

      terminator ->
        new_acc =
          if len == 0 do
            acc
          else
            [binary_part(original, skip, len) | acc]
          end

        tokenize(rest, original, skip + len + 1, 0, [terminator | new_acc])
    end
  end

  defp get_terminator(character) do
    case character do
      "|" -> @field
      "~" -> @repetition
      "^" -> @component
      "&" -> @sub_component
      "\r" -> @carriage_return
      "\n" -> @carriage_return
      _ -> nil
    end
  end
```

Comment (DarkMarmot) [reply lib/hl7/parser.ex:13]: that looks really nice!!! :) does it retain performance that way?

Comment (calvinb) [reply lib/hl7/parser.ex:13]: Yes, it does! I need to make Benchee a regular habit. Performance metrics almost always defy my intuitions.

File: lib/hl7/parser.ex (hunk 1 above)
Comment (DarkMarmot) [reply lib/hl7/parser.ex]: rewritten to be much faster (new grossness!)

Comment (DarkMarmot) [reply lib/hl7/parser.ex]: ps - the way it uses the data here to cut strings is identical to (ie stolen from) Jason. It's apparently a perf thing the way data is stored and handled.

File: lib/hl7/parser.ex (hunk 6 above)
Comment (DarkMarmot) [reply lib/hl7/parser.ex]: alas, i've found that unrolling things even farther gave me _big_ speed boosts so I have to table this :(

File: lib/hl7/parser.ex (hunk 5 above)
Comment (DarkMarmot) [reply lib/hl7/parser.ex]: very cool! but alas, it slows things down by 20% :(

File: lib/hl7/parser.ex [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    tokenize(text, original, skip + len + 1, 0, [terminator, string | acc])
+  end
+
+  defp to_lists(tokens) do
+    split_by(tokens, @carriage_return)
+    |> Enum.reject(&(&1 == []))
+    |> Enum.map(&to_segment/1)
+  end
+
+  defp to_segment(tokens) do
```
[the remark this answers is not in the document]
Comment (DarkMarmot) [reply lib/hl7/parser.ex]: ok, this is brilliant @calvinb ! i _really_ want to put it in... but am torn because it does seem to incur a small perf penalty rather than inlining every option :(

[the remark this answers is not in the document]
Comment (calvinb) [reply lib/hl7/parser.ex]: I live to slow things down.

File: lib/hl7/message.ex [hunk 8]
```diff
@@ -124,15 +124,16 @@ defmodule HL7.Message do
...
-    %HL7.Message{segments: parsed_segments, fragments: fragments, header: header}
+  def new(%HL7.RawMessage{raw: raw_text}) do
+    #    {segments, fragments} =
+    #      raw_text
+    #      |> String.split(@segment_terminator, trim: true)
+    #      |> Enum.split_with(&has_segment_name(&1, header.separators.field))
+    #
+    #    parsed_segments = segments |> Enum.map(&split_segment_text(&1, header.separators))
+    #
+    #    %HL7.Message{segments: parsed_segments, fragments: fragments, header: header}
```
Comment (calvinb) [inline lib/hl7/message.ex:135]: 🐓🪵?

File: lib/hl7/separators.ex [hunk 9]
```diff
@@ -6,7 +6,7 @@ defmodule HL7.Separators do
   """
 
   # default HL7 separators
-  # |^~\&
+  # |^~\&#
```
Comment (calvinb) [inline lib/hl7/separators.ex:9]: Is this intentional?

File: lib/hl7/message.ex (hunk 8 above)
Comment (DarkMarmot) [reply lib/hl7/message.ex:135]: doh yes! how do you put emoji in this easily? i don't see a button for it :(

File: lib/hl7/dynamic_tokenizer.ex [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  defp tokenize_terminator(text, original, skip, 0, acc, terminator, separators) do
+    tokenize(text, original, skip + 1, 0, [terminator | acc], separators)
+  end
+
+  defp tokenize_terminator(text, original, skip, len, acc, terminator, separators) do
+    string = binary_part(original, skip, len)
+    tokenize(text, original, skip + len + 1, 0, [terminator, string | acc], separators)
+  end
+
+  @compile {:inline, tokenize: 6, tokenize_terminator: 7}
```
Comment (calvinb) [inline lib/hl7/dynamic_tokenizer.ex:103]: 🎩

Comment (DarkMarmot) [reply lib/hl7/dynamic_tokenizer.ex:103]: 🐰

File: lib/hl7/message.ex (hunk 8 above)
Comment (calvinb) [reply lib/hl7/message.ex]: ctrl+cmd+space

File: lib/hl7/separators.ex (hunk 9 above)
Comment (calvinb) [reply lib/hl7/separators.ex:9]: TIL
